src/core/shards/network.cpp

```diff
@@ -254,45 +274,106 @@ struct Server : public NetworkBase {
-                                      kcp = it->second.kcp;
+  boost::lockfree::queue<const SHWire *> _stopWireQueue{16};
 
-                                      lock.unlock();
-                                    }
+  void wireOnStop(const SHWire::OnStopEvent &e) {
+    SHLOG_TRACE("Wire {} stopped", e.wire->name);
 
-                                    ikcp_input(kcp, (char *)recv_buffer.data(), bytes_recvd);
+    _stopWireQueue.push(e.wire);
+  }
 
-                                    // keep receiving
-                                    do_receive();
-                                  }
-                                });
+  void do_receive() {
+    thread_local std::vector<uint8_t> recv_buffer(0xFFFF);
+    _socket->async_receive_from(
+        boost::asio::buffer(recv_buffer.data(), recv_buffer.size()), _sender,
+        [this](boost::system::error_code ec, std::size_t bytes_recvd) {
+          if (!ec && bytes_recvd > 0) {
+            ikcpcb *kcp = nullptr;
+            std::shared_lock<std::shared_mutex> lock(peersMutex);
+            auto it = _end2Peer.find(_sender);
+            if (it == _end2Peer.end()) {
+              // new peer
+              lock.unlock();
+
+              // we write so hard lock this
+              std::unique_lock<std::shared_mutex> lock(peersMutex);
+
+              // new peer
+              try {
+                auto peer = _pool->acquire(_composer, (void *)0);
+                _end2Peer[_sender] = peer;
+                peer->endpoint = _sender;
+                peer->user = this;
+                peer->kcp->user = peer.get();
+                peer->kcp->output = &Server::udp_output;
+                SHLOG_DEBUG("Added new peer: {} port: {}", peer->endpoint->address().to_string(), peer->endpoint->port());
+
+                // Assume that we recycle containers so the connection might already exist!
+                if (!peer->onStopConnection) {
+                  _wire2Peer[peer->wire.get()] = peer;
+                  peer->onStopConnection = peer->wire->dispatcher.sink<SHWire::OnStopEvent>().connect<&Server::wireOnStop>(this);
+                }
+
+                // set wire ID, in order for Events to be properly routed
+                // for now we just use ptr as ID, until it causes problems
+                peer->wire->id = reinterpret_cast<entt::id_type>(peer.get());
+
+                kcp = peer->kcp;
+              } catch (std::exception &e) {
+                SHLOG_ERROR("Error acquiring peer: {}", e.what());
+
+                // keep receiving
+                return do_receive();
+              }
+            } else {
+              // existing peer
+              kcp = it->second->kcp;
+
+              lock.unlock();
+            }
+
+            ikcp_input(kcp, (char *)recv_buffer.data(), bytes_recvd);
+
+            // keep receiving
+            return do_receive();
+          }
+        });
   }
 
   std::vector<uint8_t> _buffer;
 
+  IterableExposedInfo _sharedCopy;
```

Comment:
Can you use `ExposedInfo` instead? I want to get rid of this weird IterableArray eventually.

This should be possible:
```cpp
_sharedCopy = ExposedInfo(data.shared);
_sharedCopy.push_back(...);
```

Reply:
but why? what you suggested would not work, I need a clone of the whole array.
Also what's weird of IterableArray, just a wrapper on top of the simple C array we have? Lets KISS?

Reply:
Btw we do this in many other places.